Fix pyright errors in `arcade.gl`

cspotcode opened the pull request:
Followup to #1751.  Will need a rebase after #1751 gets squashed.  Requesting a review from @einarf , but no rush, I know you're slowly getting back into things.

cspotcode committed:
Re-apply all the `arcade.gl` changes in a new PR

This reverts commit bfee79a186a1188f788a26155a3bf0218ea7f0b7.

gran4 commented on arcade/cache/texture.py:
```diff
@@ -91,13 +91,13 @@ def put(
         if strong:
             self._strong_entries.put(name, texture)
-            if self._strong_file_entries.get(file_path):
+            if self._strong_file_entries.get(str(file_path)):
```
Shouldn't the type be str and not Union[str, Path] if you are correct?

einarf replied:
I will redo all the cache stuff anyway, but this doesn't hurt.

gran4 commented on arcade/gl/context.py:
```diff
@@ -1,12 +1,14 @@
 import logging
+import typing
```
delete import typing

cspotcode replied:
It was actually used in this file, `typing.cast()`.  I did this because `cast` already refers to something else.  But I think I should remove that bit anyway.

gran4 commented on arcade/gl/program.py:
```diff
@@ -93,7 +93,7 @@ def __init__(
             )
 
-        shaders = [(vertex_shader, gl.GL_VERTEX_SHADER)]
+        shaders: list[tuple[str, int]] = [(vertex_shader, gl.GL_VERTEX_SHADER)]
```
Shouldn't this be List[Tuple[str, int]] or does it matter?

gran4 commented on arcade/gl/types.py:
```diff
@@ -266,7 +271,7 @@ def __init__(
             )
 
-        def zip_attrs(formats, attributes):
+        def zip_attrs(formats: list[str], attributes: Sequence[str]):
```
List not list, again, does it matter?

cspotcode replied:
Good catch, python 3.8 needs it capitalized.

gran4 commented on arcade/gui/examples/scroll_area.py:
```diff
@@ -22,12 +22,13 @@
-    canvas_size = Property(default=(300, 300))
+    scroll_x = Property[float](default=0.0)
+    scroll_y = Property[float](default=0.0)
+    canvas_size = Property[Point](default=(300.0, 300.0))
```
I think you can remove [float], and stuff. Unless there is something I am misunderstanding

gran4 commented on arcade/gui/property.py:
```diff
@@ -31,22 +29,21 @@ class Property:
         if obs is None:
-            obs = _Obs()
-            obs.value = self.default_factory(self, instance)
+            obs = _Obs(self.default_factory(self, instance))
```
I remember doing this file, we might have some conflicts

gran4 commented on arcade/gui/property.py:
```diff
@@ -77,9 +74,9 @@ def bind(self, instance, callback):
+    def __get__(self, instance, owner) -> P:
         if instance is None:
-            return self
+            return self # type: ignore
```
Make it a Union[P, Property]?

gran4 commented on arcade/gui/widgets/layout.py:
```diff
@@ -662,10 +662,10 @@ def do_layout(self):
         ]
-        max_height_per_row = [
+        max_height_per_row: list[list[tuple[float, int]]] = [
             [(0, 1) for _ in range(self.column_count)] for _ in range(self.row_count)
```
List[List[Tuple[int, int]]] instead of list[list[tuple[int, int]]]

gran4 commented on arcade/gui/widgets/slider.py:
```diff
@@ -226,7 +226,7 @@ def do_render(self, surface: Surface):
 
     def _cursor_pos(self) -> Tuple[int, int]:
-        return self.value_x, self.y + self.height // 2
+        return self.value_x, int(self.y + self.height // 2)
```
value_x looks like a float, so instead of making self.y + self.height // 2 an int, change -> Tuple[int, int]: to -> Tuple[float, float]:

gran4 commented on arcade/texture/transforms.py:
```diff
@@ -217,7 +217,7 @@ def transform_hit_box_points(
 # Key is the vertex order
 # Value is the orientation (flip_left_right, flip_top_down, rotation)
-ORIENTATIONS = {
+ORIENTATIONS: dict[tuple[int, int, int, int], tuple[int, bool, bool]] = {
```
Dict[Tuple[]]

cspotcode commented:
@gran4 Thanks for taking a look, but unfortunately, this PR can't be reviewed yet.  It's stacked on top of #1751 so the diff is all wonky.  It'll be reviewable once #1751 merges.

gran4 commented:
Oh, ok you should make it a draft then. (unless there is something I don't understand)

einarf commented:
The gl stuff looks perfectly reasonable 👍

cspotcode commented:
@einarf Thanks for taking a look!  I marked this PR as draft while we wait for #1751 to merge.  Once #1751 is merged, I'll rebase this PR, bring it out of draft, and ping you for a merge.

cspotcode committed:
Address review feedback

cspotcode committed:
Fix typecheck error

cspotcode commented:
I rebased yesterday, but there are still one or two mypy errors I need to understand and fix before this can merge.  I'll solicit a review after I've fixed them.

cspotcode committed:
Merge remote-tracking branch 'origin/development' into fix-pyright-errors-arcade-gl

cspotcode commented:
@einarf I need some help understanding some of the gl types.

https://github.com/pythonarcade/arcade/blob/889772df9b9c94d77bbb09a500e9dcb0be180e53/arcade/gl/vertex_array.py#L195-L196
I get a typecheck diagnostic `Argument 1 to "gl_name" has incompatible type "Optional[int]"; expected "int"`
This is because `prog_attr.gl_type` is `Optional[int]`

Is that correct, should `AttribFormat.gl_type` be `Optional`?

`AttribFormat` is instantiated here:
https://github.com/pythonarcade/arcade/blob/889772df9b9c94d77bbb09a500e9dcb0be180e53/arcade/gl/program.py#L394-L401
and here:
https://github.com/pythonarcade/arcade/blob/889772df9b9c94d77bbb09a500e9dcb0be180e53/arcade/gl/types.py#L311-L314

For the latter, `gl_type` is possibly `None` because it comes from `_formats`, which has some `None` values in x1, x2, x4, x8
https://github.com/pythonarcade/arcade/blob/889772df9b9c94d77bbb09a500e9dcb0be180e53/arcade/gl/types.py#L196-L220

cspotcode committed:
address review feedback

cspotcode committed:
Fix last type issue
